lib/combinat.gi

```diff
@@ -2498,18 +2495,130 @@ InstallGlobalFunction( IteratorOfPartitionsSet, function( s )
+          if m = iter!.sz then
+            return out;
+          else
+            part := StructuralCopy(iter!.next);
+          fi;
+        fi;
+      od;
+    end );
 
-    isDone := iter -> iter!.next = false;
+BindGlobal( "NextIterator_PartitionsSetGivenSizeOrLess", function( iter )
+      local j, max, part, m, out, i;
+      if Length(iter!.s) = 0 or Length(iter!.s)<iter!.sz then
+        iter!.next := false;
+        return [];
+      fi;
+      part := StructuralCopy(iter!.next);
+      while true do
+        j := Size(iter!.next);
+        # Compute next restricted growth string using a_i \leq Max({a_0 .. a_{i-1}})+1
+        while j > 1 do
+          max := Minimum(Maximum(iter!.next{[1 .. j-1]}), iter!.sz-2);
+          if iter!.next[j] <= max then
+            iter!.next[j] := iter!.next[j]+1;
+            m := Maximum(iter!.next)+1;
+            break;
+          else
+            iter!.next[j] := 0;
+            j := j-1;
+          fi;
+        od;
+        m := Maximum(part)+1;
+        # Convert restricted growth string to partition of set
+        out := List([1 .. m], t -> []);
+        for i in [1 .. Size(part)] do
+          Add(out[part[i]+1], iter!.s[i]);
+        od;
+        # this is the final iteration if in the next iteration we cycle through to the start again
+        if ForAll([2 .. Size(iter!.next)], t -> iter!.next[t]=0) then
+          iter!.next := false;
+          return out;
+        fi;
+        return out;
+      od;
+    end );
 
-    return IteratorByFunctions( rec(
-      IsDoneIterator := isDone,
-      NextIterator := nextIterator,
-      ShallowCopy := shallowCopy,
-      s := Immutable(s),
-      next := ListWithIdenticalEntries(Size(s), 0) ) );
+BindGlobal( "ShallowCopy_PartitionsSet",
+    iter -> rec( next := ShallowCopy( iter!.next ), s := iter!.s, sz := iter!.sz ) );
+
+InstallGlobalFunction( IteratorOfPartitionsSet , function( s, arg... )
+    local k, r;
+
+    if not IsSet(s) then
+      Error( "<s> must be a set" );
+    fi;
+
+    r := rec(
+            IsDoneIterator := IsDoneIterator_PartitionsSet,
+            NextIterator := NextIterator_PartitionsSet,
+            ShallowCopy := ShallowCopy_PartitionsSet,
+            s := Immutable(s),
+            next := ListWithIdenticalEntries(Size(s), 0),
+            sz := fail,
+         );
+
+    if Length( arg ) = 1 or Length( arg ) = 2 then
+      k := arg[1];
+      if not IsInt(k) then
+        Error("usage: <k> must be an integer");
+      elif k<1 or k > Length(s) then
+        Error("usage: <k> must be between 1 and size of <s>");
+      fi;
+      r.sz := k;
+      r.NextIterator := NextIterator_PartitionsSetGivenSize;
+
+      if Length( arg ) = 2 then
+        if arg[2] = true then
+          r.NextIterator := NextIterator_PartitionsSetGivenSizeOrLess;
+        elif arg[2] <> false then
+          Error("usage: <flag> must be true or false");
+        fi;
+      fi;
+    elif Length( arg ) = 1 or Length( arg ) = 2 then
```

Comment:
```suggestion
    elif Length( arg ) > 2 then
```
@jesselansdown I think that's the change you are suggesting, right? I agree with it, then

Reply:
Yep, that's the one!